lnwallet/channel.go

```diff
@@ -565,40 +565,189 @@ type htlcView struct {
 	theirUpdates []*PaymentDescriptor
 }
 
-// fetchHTLCView returns all the candidate HTLC updates which should be
-// considered for inclusion within a commitment based on the passed HTLC log
-// indexes.
-func (lc *LightningChannel) fetchHTLCView(theirLogIndex, ourLogIndex uint32) *htlcView {
+// getHTLCView convert HTLC logs into htlcView representation.
+func (lc *LightningChannel) getHTLCView() *htlcView {
 	var ourHTLCs []*PaymentDescriptor
 	for e := lc.ourUpdateLog.Front(); e != nil; e = e.Next() {
 		htlc := e.Value.(*PaymentDescriptor)
+		ourHTLCs = append(ourHTLCs, htlc)
+	}
+
+	var theirHTLCs []*PaymentDescriptor
+	for e := lc.theirUpdateLog.Front(); e != nil; e = e.Next() {
+		htlc := e.Value.(*PaymentDescriptor)
+		theirHTLCs = append(theirHTLCs, htlc)
+	}
+
+	return &htlcView{
+		ourUpdates:   ourHTLCs,
+		theirUpdates: theirHTLCs,
+	}
+}
+
+type EntryHandler func(entry *PaymentDescriptor,
+	isIncoming, isAddEntry bool) (bool, error)
+
+// processHTLCView processes all update entries in both HTLC update logs,
+// producing the new view which is the result of applying processing
+// function to the HTLC view set. Serves as "map", "filter" function
+// for htlcView set.
+func (lc *LightningChannel) processHTLCView(view *htlcView,
+	handler EntryHandler) (*htlcView, error) {
+
+	newView := &htlcView{}
+
+	// We use two maps, one for the local log and one for the remote log to
+	// keep track of which entries we need to skip. We skip an entry
+	// whenever we find a settle or a timeout modifying an entry.
+	skipUs := make(map[uint32]struct{})
+	skipThem := make(map[uint32]struct{})
+
+	// First we run through non-add entries in both logs, populating the
+	// skip sets and mutating the current chain state by applying handler
+	// function.
+	for _, entry := range view.ourUpdates {
+		if entry.EntryType == Add {
+			continue
+		}
+
+		addEntry := lc.theirLogIndex[entry.ParentIndex].Value.(*PaymentDescriptor)
+		skipThem[addEntry.Index] = struct{}{}
+
+		if include, err := handler(entry, true, false); err != nil {
```

Comment:
It looks like here, and directly, below the second bool argument should be flipped. When processing entries within our log, they're _outgoing_ modifications, not incoming.